java/compression/src/main/java/org/apache/arrow/compression/Lz4CompressionCodec.java

```diff
@@ -38,109 +36,39 @@
+  protected ArrowBuf doCompress(BufferAllocator allocator, ArrowBuf uncompressedBuffer) {
     byte[] inBytes = new byte[(int) uncompressedBuffer.writerIndex()];
     PlatformDependent.copyMemory(uncompressedBuffer.memoryAddress(), inBytes, 0, uncompressedBuffer.writerIndex());
     ByteArrayOutputStream baos = new ByteArrayOutputStream();
     try (InputStream in = new ByteArrayInputStream(inBytes);
          OutputStream out = new FramedLZ4CompressorOutputStream(baos)) {
       IOUtils.copy(in, out);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
     }
 
     byte[] outBytes = baos.toByteArray();
 
     ArrowBuf compressedBuffer = allocator.buffer(CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH + outBytes.length);
-
-    long uncompressedLength = uncompressedBuffer.writerIndex();
-    if (!MemoryUtil.LITTLE_ENDIAN) {
-      uncompressedLength = Long.reverseBytes(uncompressedLength);
-    }
-    // first 8 bytes reserved for uncompressed length, according to the specification
-    compressedBuffer.setLong(0, uncompressedLength);
-
     PlatformDependent.copyMemory(
         outBytes, 0, compressedBuffer.memoryAddress() + CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH, outBytes.length);
     compressedBuffer.writerIndex(CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH + outBytes.length);
     return compressedBuffer;
   }
 
-  @Override
-  public ArrowBuf decompress(BufferAllocator allocator, ArrowBuf compressedBuffer) {
-    Preconditions.checkArgument(compressedBuffer.writerIndex() <= Integer.MAX_VALUE,
-        "The compressed buffer size exceeds the integer limit");
-
-    Preconditions.checkArgument(compressedBuffer.writerIndex() >= CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH,
-        "Not enough data to decompress.");
-
-    long decompressedLength = compressedBuffer.getLong(0);
-    if (!MemoryUtil.LITTLE_ENDIAN) {
-      decompressedLength = Long.reverseBytes(decompressedLength);
-    }
-
-    if (decompressedLength == 0L) {
-      // shortcut for empty buffer
-      compressedBuffer.close();
-      return allocator.getEmpty();
-    }
-
-    if (decompressedLength == CompressionUtil.NO_COMPRESSION_LENGTH) {
-      // no compression
-      return CompressionUtil.extractUncompressedBuffer(compressedBuffer);
-    }
-
-    try {
-      ArrowBuf decompressedBuffer = doDecompress(allocator, compressedBuffer);
-      compressedBuffer.close();
-      return decompressedBuffer;
-    } catch (IOException e) {
-      throw new RuntimeException(e);
-    }
-  }
-
-  private ArrowBuf doDecompress(BufferAllocator allocator, ArrowBuf compressedBuffer) throws IOException {
-    long decompressedLength = compressedBuffer.getLong(0);
-    if (!MemoryUtil.LITTLE_ENDIAN) {
-      decompressedLength = Long.reverseBytes(decompressedLength);
-    }
+  protected ArrowBuf doDecompress(BufferAllocator allocator, ArrowBuf compressedBuffer) {
+    long decompressedLength = readUncompressedLength(compressedBuffer);
```

Comment:
i made a suggestion for a slightly different signature here that would eliminated the need to reference methods on the parent class.  If we go in that direction, do you think it might make sense to use composition instead of inheritance?  (haven't thought about it deeply).

Reply:
Here we are applying the template method design pattern, so inheritance seems a natural choice?